platform/core/src/services/HangingProtocolService/HangingProtocolService.ts

```diff
@@ -619,46 +633,61 @@ export default class HangingProtocolService extends PubSubService {
 
     // check if displaySetSelectors are used by other viewports, and
     // store the viewportIndex and displaySetInstanceUIDs that need to be updated
-
     const viewportsToUpdate = [];
     protocolViewports.forEach((viewport, index) => {
-      let viewportNeedsUpdate;
-      for (const displaySet of viewport.displaySets) {
-        if (
+      const viewportNeedsUpdate = viewport.displaySets.some(
+        displaySet =>
           displaySet.id === displaySetSelectorId &&
           (displaySet.matchedDisplaySetsIndex || 0) === matchedDisplaySetsIndex
-        ) {
-          viewportNeedsUpdate = true;
-          break;
-        }
-      }
+      );
 
       if (viewportNeedsUpdate) {
-        // we can then loop over the displaySets and choose all of them,
-        // but for the one that matches the oldDisplaySetInstanceUID we need to
-        // replace it with the newDisplaySetInstanceUID
-        const {
-          displaySetInstanceUIDs,
-          displaySetOptions,
-        } = viewport.displaySets.reduce(
-          (acc, displaySet) => {
-            const { id } = displaySet;
+        // Try to recompute the viewport options based on the current
+        // viewportIndex that needs update but from its old/original un-computed
+        // viewport & displaySet options
+        if (originalProtocolStage) {
+          const originalViewport = originalProtocolStage.viewports[index];
+          const originalViewportOptions = originalViewport.viewportOptions;
+          const originalDisplaySetOptions = originalViewport.displaySets;
+
+          viewport.viewportOptions = this.getComputedOptions(
+            JSON.parse(JSON.stringify(originalViewportOptions)),
+            [newDisplaySetInstanceUID]
+          );
 
-            let {
-              displaySetInstanceUID: displaySetInstanceUIDToUse,
-            } = this.displaySetMatchDetails.get(id);
+          const recomputedDisplaySetOptions = this.getComputedOptions(
+            JSON.parse(JSON.stringify(originalDisplaySetOptions)),
+            [newDisplaySetInstanceUID]
+          );
 
-            if (displaySet.id === displaySetSelectorId) {
-              displaySetInstanceUIDToUse = newDisplaySetInstanceUID;
-            }
+          if (recomputedDisplaySetOptions) {
+            viewport.displaySets = viewport.displaySets.map(
+              (displaySet, index) => {
+                return {
+                  ...displaySet,
+                  ...recomputedDisplaySetOptions[index],
+                };
+              }
+            );
+          }
+        }
 
-            acc.displaySetInstanceUIDs.push(displaySetInstanceUIDToUse);
-            acc.displaySetOptions.push(displaySet);
+        const displaySetInstanceUIDs = [];
+        const displaySetOptions = [];
 
-            return acc;
-          },
-          { displaySetInstanceUIDs: [], displaySetOptions: [] }
-        );
+        for (const prevDisplaySet of viewport.displaySets) {
```

Comment:
Can you drop this into a function - just that the code is already pretty long, and making it into a single line makes the whole function easier to understand.